crypto/src/asset/amount.rs

```diff
@@ -30,6 +34,34 @@ impl Amount {
     }
 }
 
+#[derive(Clone)]
+pub struct AmountVar {
+    pub amount: FqVar,
+}
+
+impl AllocVar<Amount, Fq> for AmountVar {
+    fn new_variable<T: std::borrow::Borrow<Amount>>(
+        cs: impl Into<ark_relations::r1cs::Namespace<Fq>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: ark_r1cs_std::prelude::AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+        let amount1 = f()?;
+        let amount: Amount = *amount1.borrow();
```

Comment:
nit: since we're cloning the value, could this be
```suggestion
        let amount: Amount = *f()?.borrow();
```